ee/hogai/graph/insights/nodes.py

```diff
@@ -242,62 +307,254 @@ def _format_insights_page(self, page_number: int) -> str:
 
         return "\n".join(formatted_insights)
 
-    def _parse_insight_ids(self, response_content: str) -> list[int]:
-        """Parse insight IDs from LLM response."""
-        import re
+    def _get_all_loaded_insight_ids(self) -> set[int]:
+        """Get all insight IDs from loaded pages."""
+        all_ids = set()
+        for page_insights in self._loaded_pages.values():
+            for insight in page_insights:
+                all_ids.add(insight.id)
+        return all_ids
+
+    def _find_insight_by_id(self, insight_id: int) -> Insight | None:
```

Comment:
It seems like O(n2) complexity and is being called multiple times. Would it make sense to cache it somehow, or are the loaded pages too dynamic? Or maybe just build the `self` index dictionary on the go? With thousands of insights it can get a bit expensive, as I see it.